onnxruntime/core/mlas/lib/sqnbitgemm.h

```diff
@@ -0,0 +1,299 @@
+{
+    return BlkLen * BlkBitWidth / 8;
+}
+
+template <size_t BlkBitWidth>
+constexpr MLAS_FORCEINLINE size_t
+MlasQNBitZeroPointsForBlksSizeInBytes(size_t BlkCount)
+{
+    if constexpr (BlkBitWidth <= 4) {
+        return MlasDivRoundup(BlkCount, 2);  // 2 blocks per byte
+    } else {
+        return BlkCount;
+    }
+}
+
+MLAS_FORCEINLINE void
+MlasAddBiasForGemm(const float* Bias, float* C, size_t CountM, size_t CountN, size_t ldc)
+{
+    for (size_t m = 0; m < CountM; m++) {
+        const float* bias = Bias;
+        float* sum = C;
+        for (size_t n = 0; n < CountN; n += 4) {
+            if (CountN - n < 4) {
+                for (size_t nn = n; nn < CountN; nn++) {
+                    *sum += *bias;
+                    sum++;
+                    bias++;
+                }
+                break;
+            }
+
+            MLAS_FLOAT32X4 acc_x = MlasLoadFloat32x4(sum);
+            acc_x = MlasAddFloat32x4(acc_x, MlasLoadFloat32x4(bias));
+            MlasStoreFloat32x4(sum, acc_x);
+            bias += 4;
+            sum += 4;
+        }
+        C += ldc;
+    }
+}
+
+template <size_t BlkBitWidth, size_t BlkLen, typename KernelType>
+MLAS_FORCEINLINE void MLASCALL
+MlasSQNBitGemmOperation(
+    const size_t K,
+    const MLAS_SQNBIT_GEMM_DATA_PARAMS* const DataParams,
+    const size_t RangeStartM,
+    const size_t RangeCountM,
+    const size_t RangeStartN,
+    const size_t RangeCountN
+)
+{
+    const size_t lda = DataParams->lda;
+    const size_t ldc = DataParams->ldc;
+
+    const size_t k_blks = MlasDivRoundup(K, BlkLen);
+    const size_t ldb = k_blks * MlasQNBitBlkDataSizeInBytes(BlkBitWidth, BlkLen);
+    const float* A = DataParams->A + RangeStartM * lda;
+    const uint8_t* QuantBData = static_cast<const uint8_t*>(DataParams->QuantBData);
+    const float* QuantBScale = DataParams->QuantBScale;
+    const uint8_t* QuantBZeroPoint = static_cast<const uint8_t*>(DataParams->QuantBZeroPoint);
+    float* C = DataParams->C + RangeStartM * ldc + RangeStartN;
+    const float* Bias = DataParams->Bias;
+
+    if (RangeCountM == 1) {
+        size_t CountN;
+        for (size_t n = 0; n < RangeCountN; n += CountN) {
+            CountN = std::min(RangeCountN - n, (size_t)128);
+
+            //
+            // Step through each slice of matrix A along the M dimension.
+            //
+            const float* bias = (Bias == nullptr) ? nullptr : Bias + RangeStartN + n;
+            const uint8_t* b_col = QuantBData + (RangeStartN + n) * ldb;
+            const float* b_col_scale = QuantBScale + (RangeStartN + n) * k_blks;
+            const uint8_t* b_col_zp =
+                (QuantBZeroPoint == nullptr)
+                    ? nullptr
+                    : QuantBZeroPoint + (RangeStartN + n) * MlasQNBitZeroPointsForBlksSizeInBytes<BlkBitWidth>(k_blks);
+            float* c_blk = C + n;
+            const float* a_row = A;
+
+            size_t RowsRemaining = RangeCountM;
+            while (RowsRemaining > 0) {
```

Comment:
>while (RowsRemaining > 0) { [](http://example.com/codeflow?start=12&length=27)

nit: no need to use while because RangeCountM is only 1

Reply:
I followed MlasQ4GemmOperation in q4gemm.h as an example. I'm not sure if the intent of this path is to only ever run for M=1 or whether we want to handle larger M with it in the future. If it's the former, we can simplify some things including this `while` loop.

Reply:
you can remove it. I simply forgot to remove it in q4gemm.h

Reply:
updated kernel name and signature to make it clear it is for M=1 and removed loop